controllers/component_dependency_update_controller.go

```diff
@@ -0,0 +1,484 @@
+	// find the image and digest we want to update to
+	image := ""
+	tag := ""
+	digest := ""
+	for _, r := range pipelineRun.Status.Results {
+		if r.Name == ImageDigestParamName {
+			digest = r.Value.StringVal
+		} else if r.Name == ImageUrlParamName {
+			image = r.Value.StringVal
+		}
+	}
+	// Failure to find is a permanent error, so remove the finalizer
+	if image == "" {
+		log.Error(fmt.Errorf("unable to find %s param on PipelineRun, not performing nudge", ImageUrlParamName), "no image url result")
+		return r.removePipelineFinalizer(ctx, pipelineRun)
+	}
+	if digest == "" {
+		log.Error(fmt.Errorf("unable to find %s param on PipelineRun, not performing nudge", ImageDigestParamName), "no image digest result")
+		return r.removePipelineFinalizer(ctx, pipelineRun)
+	}
+	parts := strings.Split(image, ":")
+	if len(parts) != 2 {
+		log.Error(fmt.Errorf("invalid image name %s", image), "invalid image format")
+		return r.removePipelineFinalizer(ctx, pipelineRun)
+	}
+	tag = parts[1]
+	repo := parts[0]
+
+	components := applicationapi.ComponentList{}
+	err := r.Client.List(ctx, &components, client.InNamespace(pipelineRun.Namespace))
+	if err != nil {
+		log.Error(err, "failed to list components in namespace")
+		return ctrl.Result{}, err
+	}
+
+	retryTime := FailureRetryTime
+	immediateRetry := false
+
+	toUpdate := []applicationapi.Component{}
+
+	for i := range components.Items {
+		comp := components.Items[i]
+		if comp.Spec.Application == updatedComponent.Spec.Application && slices.Contains(updatedComponent.Spec.BuildNudgesRef, comp.Name) {
+			toUpdate = append(toUpdate, comp)
+		}
+	}
+	var nudgeErr error
+	immediateRetry, nudgeErr = r.UpdateFunction(ctx, r.Client, r.Scheme, r.EventRecorder, toUpdate, &BuildResult{BuiltImageRepository: repo, BuiltImageTag: tag, Digest: digest, Component: updatedComponent})
+	if nudgeErr != nil {
+		log.Error(nudgeErr, fmt.Sprintf("component update of as a result of a build of %s failed", updatedComponent.Name), l.Audit, "true")
+
+		if pipelineRun.Annotations == nil {
+			pipelineRun.Annotations = map[string]string{}
+		}
+		existing := pipelineRun.Annotations[FailureCountAnnotationName]
+		if existing == "" {
+			existing = "0"
+		}
+		failureCount, err := strconv.Atoi(existing)
+		if err != nil {
+			log.Error(err, "failed to pass retry count, not retrying")
+			return r.removePipelineFinalizer(ctx, pipelineRun)
+		}
+		failureCount = failureCount + 1
+
+		pipelineRun.Annotations[FailureCountAnnotationName] = strconv.Itoa(failureCount)
+
+		r.EventRecorder.Event(updatedComponent, corev1.EventTypeWarning, ComponentNudgeFailedEventType, fmt.Sprintf("component update failed as a result of a build for %s, retry %d/%d", updatedComponent.Name, failureCount, MaxAttempts))
+
+		if failureCount >= MaxAttempts {
+			// We are at the failure limit, nothing much we can do
+			log.Info("not retrying as max failure limit has been reached", l.Audit, "true")
+			return r.removePipelineFinalizer(ctx, pipelineRun)
+		}
```

Comment:
Should we maybe do this check first, before trying to do the update? Isn't it possible that the reconcile will get requeued for a PipelineRun which has already reached max retries?

Reply:
No, because the code that would update the failure count will also add the completed annotation. If you see a PipelineRun that has hit the failure count it will already be marked as done and ignored by the reconcile loop.

Reply:
Right, but `removePipelineFinalizer` can fail as well. And it might not even be that uncommon due to the number of controllers that reconcile on and update PipelineRuns?

So IIUC the operation that follows immediately after "not retrying as max failure limit has been reached" can fail and cause a retry

Reply:
Maybe putting the max retries check before the `r.UpdateFunction()` call could avoid retrying more than intended (unless the operation that sets the failure count annotation fails)

Reply:
There is nothing we can do about this, the removal of the finalizer and setting of the update count is the same update operation. If it does fail and a retry happens more than expected that is not a huge deal, especially given that failure will be very rare in normal circumstances. Failure in this case is failure to launch the renovate job, we don't reconcile on jobs and attempt to check the result.